/**
 * @file dialogaddtransition.h
 * @author Eliška Křeménková (xkremee00)
 * @brief Declaration of the DialogAddTransition class that handles adding new transition to the model
 * @date 2025-05-09
 */

#ifndef DIALOGADDTRANSITION_H
#define DIALOGADDTRANSITION_H

#include <QDialog>

namespace Ui {
class DialogAddTransition;
}

/**
 * @class DialogAddTransition
 * @brief A dialog window for creating and configuring a new transition.
 *
 * This class provides a simple UI for entering
 * - source and target state name
 * - list of input conditions
 * - timeout
 * 
 * It is used as a modal dialog and returns user input
 * through getter methods after acceptance.
 */
class DialogAddTransition : public QDialog
{
    Q_OBJECT

public:
    /**
     * @brief Construct a new Dialog Add Transition object
     * @param parent Parent widget, default is nullptr
     */
    explicit DialogAddTransition(QWidget *parent = 0);

    /**
     * @brief Destroy the Dialog Add Transition object
     */
    ~DialogAddTransition();

    /**
     * @brief Set available states to the dialog's combo boxes
     * @param stateNames List of available state names
     */
    void setAvailableStates(const QStringList& stateNames);

    /**
     * @brief Set available states to the dialog's combo boxes
     * @param pointers List of available state names
     */
    void setInputPointers(const QStringList& pointers);

    /**
     * @brief Get the transition source state name from the dialog
     * @return name
     */
    QString getFromState() const;

    /**
     * @brief Get the transition target state name from the dialog
     * @return name
     */
    QString getToState() const;

    /**
     * @brief Get the input conditions from the dialog
     * @return list of conditions stored as string pairs
     */
    QList<QPair<QString, QString> > getInputConditions() const;

    /**
     * @brief Get the timeout for of the transition from the dialog
     * @return timeout in milliseconds
     */
    int getTimeout() const;

private slots:
    /**
     * @brief Add a new input condition to the list of input conditions inside the dialog
     * 
     * Triggered when the user clicks the "Add" button
     * and adds data from input fields (e.g., input symbol and pointer) to the list
     */
    void on_pushButton_add_condition_clicked();

    /**
     * @brief Remove selected input condition from the list of input conditions inside the dialog
     * 
     * Triggered when the user clicks the "Remove" button
     * and removes selected condition from the list
     */
    void on_pushButton_remove_condition_clicked();

private:
    /**
     * @brief Pointer to the UI elements defined in the dialogaddtransition.ui file.
     * 
     * This is automatically generated by Qt's user interface compiler (uic) and provides
     * access to all widgets and layout elements created in the Qt Designer for DialogAddTransition.
     */
    Ui::DialogAddTransition *ui;
};

#endif // DIALOGADDTRANSITION_H
